src/mono/wasm/runtime/jiterpreter.ts

```diff
@@ -1273,6 +1290,25 @@ function generate_wasm_body (
                 builder.endBlock();
                 break;
 
+            case MintOpcode.MINT_MONO_CMPXCHG_I4:
+                builder.local("pLocals");
+                append_ldloc(builder, getArgU16(ip, 2), WasmOpcode.i32_load); // dest
+                append_ldloc(builder, getArgU16(ip, 3), WasmOpcode.i32_load); // newVal
+                append_ldloc(builder, getArgU16(ip, 4), WasmOpcode.i32_load); // expected
+                builder.callImport("cmpxchg_i32");
+                append_stloc_tail(builder, getArgU16(ip, 1), WasmOpcode.i32_store);
+                break;
+            case MintOpcode.MINT_MONO_CMPXCHG_I8:
+                // because i64 values can't pass through JS cleanly (c.f getRawCwrap and
```

Comment:
None of these values enter/exit JS since the implementation of cmpxchg_i64 is C and the trace is pure wasm, so it should be fine to use i64s directly here

Reply:
Because you're using getRawCwrap there's no JS wrapper involved (see https://github.com/dotnet/runtime/blob/98514de67e76f25b8b7b7965b27a29c2efecb620/src/mono/wasm/runtime/jiterpreter-support.ts#L931 ) - it should be a raw wasm export which is valid to shove into the import table of the trace module and wire everything through, I believe. If it's not possible it *should* be

Reply:
No, it doesn't work - there's Wasm.instaniate throws an import error. because `EMSCRIPTEN_KEEPALIVE` makes some weird signature for `mono_jiterp_cas_i64` if I pass int64_t arguments by value

Reply:
It's possible that something else was wrong when I was doing it by value.  I'm not too inclined to investigate it right now to be honest.  We could make an issue to follow-up.